src/NServiceBus.Core/Persistence/InMemory/Gateway/InMemoryGatewayDeduplication.cs

```diff
@@ -44,5 +51,48 @@ public Task<bool> DeduplicateMessage(string clientId, DateTime timeReceived, Con
         internal readonly int maxSize;
         readonly LinkedList<string> clientIdList = new LinkedList<string>();
         readonly Dictionary<string, LinkedListNode<string>> clientIdSet = new Dictionary<string, LinkedListNode<string>>();
+
+        class EnlistmentNotification : IEnlistmentNotification
+        {
+            public EnlistmentNotification(Dictionary<string, LinkedListNode<string>> clientIdSet, string clientId)
+            {
+                this.clientIdSet = clientIdSet;
+                this.clientId = clientId;
+            }
+
+            public void Prepare(PreparingEnlistment preparingEnlistment)
+            {
+                try
+                {
+                    preparingEnlistment.Prepared();
+                }
+                catch (Exception ex)
+                {
+                    preparingEnlistment.ForceRollback(ex);
+                }
+            }
+
+            public void Commit(Enlistment enlistment)
+            {
+                enlistment.Done();
+            }
+
+            public void Rollback(Enlistment enlistment)
+            {
+                lock (clientIdSet)
+                {
+                    clientIdSet.Remove(clientId);
```

Comment:
do we need to protect this operation with the lock as well?

Reply:
what's the impact of not removing the id from the IdList on the rollback scenario?